rfc: Crash-Consistent Layer Map Updates By Leveraging index_part.json (#5086)
neondatabase/neon: docs/rfcs/027-crash-consistent-layer-map-through-index-part.md

### The Problem

There are are cases where we need to make an atomic update to the layer map that involves **more than one layer**.
The best example is compaction, where we need to insert the L1 layers generated from the L0 layers, and remove the L0 layers.
As stated above, making the update to the layer map in atomic way is trivial.
But, there is no system call API to make an atomic update to a directory that involves more than one file rename and deletion.
Currently, we issue the system calls one by one and hope we don't crash.

What happens if we crash and restart in the middle of that system call sequence?
We will reconstruct the layer map according to the reconciliation process, taking as input whatever transitory state the timeline directory ended up in.

We cannot roll back or complete the timeline directory update during which we crashed, because we keep no record of the changes we plan to make.

### Problem's Implications For Comapction

The implications of the above are primarily problematic for compaction.
Specifically, the part of it that compats L0 layers into L1 layers.

> * Recap: compaction takes a set of L0 layers and reshuffels the delta records in them into L1 layer files.

Reviewer: ```suggestion
* Recap: compaction takes a set of L0 layers and reshuffles the delta records in them into L1 layer files.
```

If we crash after the index part PUT, we leak the to-be-DELETEd layer files to S3.
Leaking is fine, it's a pre-existing condition and not addressed in this RFC.

Multi-object changes that previously created and removed files in timeline dir are now atomic because the layer map updates are atomic and crash consistent:
* atomic layer map update at runtime, currently by using an RwLock in write mode
* atomic `index_part.json` update in S3, as per guarantee that S3 PUT is atomic
* local timeline dir state:
  * irrelevant for layer map content => irrelevant for atomic updates / crash consistency
  * if we crash after index part PUT, local layer files will be used, so, no on-demand downloads neede for them
  * if we crash before index part PUT, local layer files will be deleted

## Trade-Offs

`disk_consistent_lsn` can go backwards across restarts if we crash before we've finished the index part PUT.
Nobody should care about it, because the only thing that matters is `remote_consistent_lsn`.
Compute certainly doesn't care about `disk_consistent_lsn`.

If we crash before finishing the index part PUT, we have to re-do the work during restart:

> * wal ingest: losing not-yet-uploaded L0s; load on the **safekeepers** + work for pageserver

Reviewer 2: Can you sketch out how large this data is in practice?

IIUC, it is already the case that we do not recover ephemeral files on startup, so we already "reset" our WAL consumption to whatever the most recent DeltaLayer/ImageLayer had.

After the change in this RFC, is the extra WAL replay just equivalent to however deep our queue of S3 uploads was (i.e. seconds * bandwidth)?

Author: Yes exactly, I think I make that point in the next paragraph. Will restructure this section to make this clear upfront.

> * compaction: we lose the entire compaction iteration work; need to re-do it again

Reviewer 2: Currently, compaction can be very long running -- is re-doing from start unavoidable, or could we avoid totally re-doing if we modify the compaction code to do some mid-compaction uploads between key ranges?

Author: Resumable / incremental compaction is issue https://github.com/neondatabase/neon/issues/4580
I don't know Chi's compaction work well enough to tell whether he solved that as part of his prototype.

---

In this prototype, re-doing the compaction work is unavoidable.

How long running is it in practice, though? Let me see whether we have useful metrics for that.

Reviewer 2: When we tried a 1TiB example the other week, it was hours (although that was including some restarts since it was in staging)

Author: So, practically speaking, here are some approximate numbers from prod (we should fix the histogram to not count no-op iterations): https://neonprod.grafana.net/d/d9247703-8c96-44f0-b8d1-cbd4ee17062a/compaction-timings?orgId=1&from=1692706093728&to=1692852222721

Author: Was this 1TiB logical size or physical size?

Reviewer 2: logical

Reviewer 3: I think it was reimaging which was taking that long instead of L0 -> L1 compaction. We just do them sequentially.

Author: I rewrote the trade-off section, moving fundamental trade-off to the front and tried to add numbers / worst-case estimations where I could.

Please re-read that section and continue discussion in a new comment thread, repeating any remarks from this thread should I have not addressed them.

A longer budget would expose tenants that are done early to a longer downtime.
A short budget would risk throwing away more work that'd have to be re-done after restart.
In the context of this RFC, killing the process would mean losing the work that hasn't made it to S3.
We can mitigate this problem as follows:
0. initially, by accepting that we need to do the work again
1. short-term, by introducing a read-only shutdown state for tenants that are fast to shut down; that state would be equivalent to the state of a tenant in hot standby / readonly mode.
2. mid term, by not restarting pageserver in place, but using [*seamless tenant migration*](https://github.com/neondatabase/neon/pull/5029) to drain a pageserver's tenants before we restart it.

## Side-Effects Of This Design

* local `metadata` is basically reduced to a cache of which timelines exist for this tenant; i.e., we can avoid a `ListObjects` requests for a tenant's timelines during tenant load.

## Limitations

Multi-object changes that span multiple timelines aren't covered.
We currently have:
1. GC (but, it doesn't require atomicity across timelines, so, it doesn't really count)

> 2. Timeline deletion (covered by `deleted_at` marker)

Reviewer: This should probably expand on why this is about *timeline* deletion, because one might think it should read *tenant* deletion, because that's what first comes to mind when thtinking about work spanning multiple timelines. I think in the meeting we mentioned dependencies between different timelines making this multi-timeline?

Author: Hm, I think I'll just drop timeline & tenant deletion here. Both of them are really not "atomic layer map changes" but outright deletions of the layer map. No need to solve these. Will explain that.

Author: f51079302e8e7cace8ee6f2353763c147f024c07

Lastly, a temporarily clogged-up upload queue (e.g. S3 is down) shouldn't immediately render ingestion unavailable.

## Related issues

- https://github.com/neondatabase/neon/issues/4749
- https://github.com/neondatabase/neon/issues/4418
  - https://github.com/neondatabase/neon/pull/4422
- https://github.com/neondatabase/neon/issues/5077
- https://github.com/neondatabase/neon/issues/4088
  - (re)resolutions:
    - https://github.com/neondatabase/neon/pull/4696
    - https://github.com/neondatabase/neon/pull/4094
      - https://neondb.slack.com/archives/C033QLM5P7D/p1682519017949719

Note that the test case introduced in https://github.com/neondatabase/neon/pull/4696/files#diff-13114949d1deb49ae394405d4c49558adad91150ba8a34004133653a8a5aeb76 will produce L1s with the same logical content, but, as outlined in the last paragraph of the _Problem Statement_ section above, we don't want to make that  assumption in order to fix the problem.

## Implementation Plan

1. Remove support for `remote_storage=None`, because we now rely on the existence of an index part.

>     - The nasty part here is to fix all the tests that fiddle with the loca timeline directory.

Reviewer: ```suggestion
    - The nasty part here is to fix all the tests that fiddle with the local timeline directory.
```

# Crash-Consistent Layer Map Updates By Leveraging `index_part.json`

* Created on: Aug 23, 2023
* Author: Christian Schwarz

## Summary

This RFC describes a simple scheme to make layer map updates crash consistent by leveraging the `index_part.json` in remote storage.
Without such a mechanism, crashes can induce certain edge cases in which broadly held assumptions about system invariants don't hold.

## Motivation

### Background

We can currently easily make complex, atomic updates to the layer map by means of an RwLock.
If we crash or restart pageserver, we reconstruct the layer map from:
1. local timeline directory contents
2. remote `index_part.json` contents.

> The function that is responsible for this is called `reconcile_with_remote()`.

Reviewer 3: These are post #5074

* For a file name that, by its name, is present locally and in the remote `index_part.json`, but where the local file has a different size (future: checksum) than the remote file, we will delete the local file and leave the remote file as a `RemoteLayer` in the layer map.

### The Problem

There are are cases where we need to make an atomic update to the layer map that involves **more than one layer**.
The best example is compaction, where we need to insert the L1 layers generated from the L0 layers, and remove the L0 layers.
As stated above, making the update to the layer map in atomic way is trivial.
But, there is no system call API to make an atomic update to a directory that involves more than one file rename and deletion.
Currently, we issue the system calls one by one and hope we don't crash.

What happens if we crash and restart in the middle of that system call sequence?
We will reconstruct the layer map according to the reconciliation process, taking as input whatever transitory state the timeline directory ended up in.

We cannot roll back or complete the timeline directory update during which we crashed, because we keep no record of the changes we plan to make.

> ### Problem's Implications For Comapction

Reviewer 3: ```suggestion
### Problem's Implications For Compaction
```

The implications of the above are primarily problematic for compaction.

> Specifically, the part of it that compats L0 layers into L1 layers.

Reviewer 3: ```suggestion
Specifically, the part of it that compacts L0 layers into L1 layers.
```

Pageserver currently treat file size as a checksum, and because it mismatches, the node B will refuse to read data from the L1.
Effectively, the data in the L1 has become inaccessible to node B.
If node B already uploaded an index part itself, all subsequent attachments will use node B's index part, and run into the same probem.

If we ever introduce checksums instead of checking just the file size, then a mismatching bit pattern (2) will cause similar problems.

In case of (1) and (2), where we know that the logical content of the layers is still the same, we can recover by manually patching the `index_part.json` of the new node to the overwritten L1's file size / checksum.

But if (3) ever happens, the logical content may be different, and, we could have truly lost data.

Given the above considerations, we should avoid making correctness of split-brain protection dependent on overwrites preserving _logical_ layer file contents.
**It is a much cleaner separation of concerns to require that layer files are truly immutable in S3, i.e., PUT once and then only DELETEd, never overwritten (overPUTted).**

## Design

> Instead of reonciling a layer map from local timeline directory contents and remote index part, this RFC proposes to view the remote index part as authoritative during timeline load.

Reviewer 3: ```suggestion
Instead of reconciling a layer map from local timeline directory contents and remote index part, this RFC proposes to view the remote index part as authoritative during timeline load.
```

Local layer files will be recognized if they match what's listed in remote index part, and removed otherwise.

During **timeline load**, the only thing that matter is the remote index part content.
Essentially, timeline load becomes much like attach, except we don't need to prefix-list the remote timelines.
The local timeline dir's `metadata` file does not matter.
The layer files in the local timeline dir are seen as a nice-to-have cache of layer files that are in the remote index part.
Any layer files in the local timeline dir that aren't in the remote index part are removed during startup.

> The `reconcile_with_remote()` no longer "merges" local timeline dir contents with the remote index part.

Reviewer 3: ```suggestion
The `Timeline::load_layer_map()` no longer "merges" local timeline dir contents with the remote index part.
```

Instead, it treats the remote index part as the authoritative layer map.
If the local timeline dir contains a layer that is in the remote index part, that's nice, and we'll re-use it if file size (and in the future, check sum) match what's stated in the index part.
If it doesn't match, we remove the file from the local timeline dir.

After load, **at runtime**, nothing changes compared to what we did before this RFC.
The procedure for single- and multi-object changes is reproduced here for reference:
* For any new layers that the change adds:
  * Write them to a temporary location.
  * While holding layer map lock:
    * Move them to the final location.
    * Insert into layer map.
* Make the S3 changes.
  We won't reproduce the remote timeline client method calls here because these are subject to change.
  Instead we reproduce the sequence of s3 changes that must result for a given single-/multi-object change:
    * PUT layer files inserted by the change.
    * PUT an index part that has insertions and deletions of the change.
    * DELETE the layer files that are deleted by the change.

>       * With the [split-brain protection RFC](https://github.com/neondatabase/neon/pull/4919), the deletions will be written to deletion queue instead of scheduled.

Reviewer 3: Or already with #4938 the deletes will come later.

Author: 1faa19b96531a1653e48eae474f7af976415c63b
